### PR TITLE
Spike: submit proposal

### PR DESCRIPTION
**Pull request by aidan-starke**
## Why

> Pin proposal justification to IPFS & submit to CENNZnet

## What is changing

 - Add `Enactment Delay` field
 - Add null address as placeholder for `Extrinsic Hash` (Advanced)
 - Update form field labels

### CODE REVIEW

File: packages/gov-app/.env.development [hunk 1]
```diff
@@ -5,3 +5,5 @@ NEXTAUTH_SECRET=
 NEXTAUTH_URL=http://localhost:4200
 TWITTER_CLIENT_ID=Sk4zcnBkZk5mVm9OczFPQUU2OXk6MTpjaQ
 TWITTER_CLIENT_SECRET=
+PINATA_JWT=
+NEXT_PUBLIC_IPFS_GATEWAY=
```
Comment (ken-futureverse) [inline packages/gov-app/.env.development]: we can fill `NEXT_PUBLIC_IPFS_GATEWAY` in for `.env.development` and `.env.production`, it's public anyway

File: packages/gov-app/libs/constants.ts [hunk 2]
```diff
@@ -41,3 +41,10 @@ export const TWITTER_CLIENT: Pick<
 	clientId: process.env.TWITTER_CLIENT_ID,
 	clientSecret: process.env.TWITTER_CLIENT_SECRET,
 };
+
+export const NULL_ADDRESS = "0x0000000000000000000000000000000000000000";
+
+export const PINATA_JWT: string = process.env.PINATA_JWT;
```
Comment (ken-futureverse) [inline packages/gov-app/libs/constants.ts]:
```suggestion
export const PINATA_JWT: string = process.env.PINATA_JWT ?? "";
```

File: packages/gov-app/libs/utils/pinProposalToIPFS.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+			method: "POST",
+			headers: {
+				"Authorization": `Bearer ${PINATA_JWT}`,
+				"Content-Type": "application/json",
+			},
+			body: data,
+		}
+	);
+
+	return response.json();
```
Comment (ken-futureverse) [inline packages/gov-app/libs/utils/pinProposalToIPFS.ts]: could do a little more error handling here, plus `response.json()` is an async operation

```suggestion
const data = await response.json();

if(!response.ok) {
  throw {code: response.status message: data?.message ?? response.statusText};
}

return data;
```

File: packages/gov-app/pages/proposals/new.tsx [hunk 4]
```diff
@@ -89,19 +127,68 @@ const NewProposal: NextPage = () => {
...
 	const [busy, setBusy] = useState<boolean>(false);
+
+	const { api } = useCENNZApi();
+	const { selectedAccount, wallet } = useCENNZWallet();
+	const signer = wallet?.signer;
+
 	const onFormSubmit: FormEventHandler<HTMLFormElement> = useCallback(
-		(event) => {
+		async (event) => {
 			event.preventDefault();
```
Comment (ken-futureverse) [inline packages/gov-app/pages/proposals/new.tsx:137]: passing the field values works but is a bit brittle as we need to worry about it when adding new fields. There is a better way to handle it, you can tap into the new `FormData` API instead, which collects all the values of all inputs that are wrapped around the `form` element

```
const data = new FormData(event.target as HTMLFormElement);
const proposalTitle = data.get("proposalTitle"); // assuming you have the field named `"proposalTitle"`
...
```

File: packages/gov-app/pages/proposals/new.tsx [hunk 5]
```diff
@@ -11,13 +12,24 @@ import {
...
 
 const NewProposal: NextPage = () => {
 	const [proposalTitle, setProposalTitle] = useState<string>("");
-	const [advancedOpen, setAdvancedOpen] = useState<boolean>();
-
-	const { busy, onFormSubmit } = useFormSubmit();
+	const [proposalDetails, setProposalDetails] = useState<string>("");
+	const [proposalExtrinsic, setProposalExtrinsic] = useState<string>("");
+	const [proposalDelay, setProposalDelay] = useState<number>(0);
+
```
Comment (ken-futureverse) [inline packages/gov-app/pages/proposals/new.tsx]: a `useControlledInput` might look quite elegant here, notice the use of the generic

```
const {value: proposalDetails, onChange: onProposalDetailsChange} = useControlledInput<string>("");

const {value: proposalDelay, onChange: onProposalDelayChange} = useControlledInput<number>(0);
```

**Review by ken-futureverse — CHANGES_REQUESTED** [review]

File: packages/gov-app/.env.development (hunk 1 above)
Comment (aidan-starke) [reply packages/gov-app/.env.development]: do we have a gateway setup for gov-services?

**Pull request merged**